plasmapy/simulation/particle_tracker/save_routines.py

```diff
@@ -0,0 +1,215 @@
+"""
+Module containing save routines for the particle tracker.
+"""
+
+__all__ = [
+    "AbstractSaveRoutine",
+    "DoNotSaveSaveRoutine",
+    "SaveOnceOnCompletion",
+    "IntervalSaveRoutine",
+]
+
+from abc import ABC, abstractmethod
+from pathlib import Path
+
+import astropy.units as u
+import h5py
+
+
+class AbstractSaveRoutine(ABC):
+    """Abstract base class containing the necessary methods for a
+    `~plasmapy.simulation.particle_tracker.particle_tracker.ParticleTracker` save routine.
+
+    The save routine class is responsible for defining the conditions and hooks
+    for saving.
+
+    Parameters
+    ----------
+    output_directory : `~pathlib.Path`, optional
+        Output for objects that are saved to disk. If a directory is not specified
+        then a memory save routine is used.
+
+
+    Notes
+    -----
+    After every push, the `post_push_hook` method is called with the
+    respective `~plasmapy.simulation.particle_tracker.particle_tracker.ParticleTracker` object passed as a parameter.
+    Then, the hook calls `save_now` to determine whether or not the simulation state should be saved.
+    """
+
+    def __init__(self, output_directory: Path | None = None) -> None:
+        self.output_directory = output_directory
+
+        self._results = {}
+        self._quantities = {
```

Comment:
Save time in AbstractSave: should be a shape[ntimes] array if the timestep is synchronized, otherwise a shape[ntimes,nparticles]. I think this should be the default since without time you can't really plot x,v. In subclasses, like the charged particle radiography subclass, you can overwrite it and not save time as it makes sense.

Reply:
Ok I just added "time" to all of the `_quantities` lists for the save routines. By default it will follow this behavior since that's how the underlying `ParticleTracker` keeps track of time anyways (one number for synchronized timestep, otherwise an array with entries for each particle)